homeassistant/components/tvoverlay/notify.py

```diff
@@ -0,0 +1,289 @@
+    hass: HomeAssistant,
+    config: ConfigType,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> TvOverlayNotificationService | None:
+    """Get the TvOverlay notification service."""
+    return TvOverlayNotificationService(discovery_info) if discovery_info else None
+
+
+class TvOverlayNotificationService(BaseNotificationService):
+    """Notification service for TvOverlay."""
+
+    def __init__(self, config: dict[str, Any]) -> None:
+        """Initialize the service."""
+        self.notify = Notifications(config[CONF_HOST])
+
+    async def _is_valid_url(self, url: str) -> bool:
+        """Check if a valid url and in allowlist_external_urls."""
+        try:
+            cv.url(url)
+        except vol.Invalid as err:
+            raise ServiceValidationError(
+                f"Invalid url '{url}'",
+                translation_domain=DOMAIN,
+                translation_key="invalid_url",
+                translation_placeholders={
+                    "url": url,
+                },
+            ) from err
+
+        if self.hass.config.is_allowed_external_url(url):
+            return True
+
+        allow_url_list = "allowlist_external_urls"
+        help_url = "https://www.home-assistant.io/docs/configuration/basic/"
+
+        raise ServiceValidationError(
+            f"Cannot send TvOverlay notification with image url '{url}' which is not secure to load data from."
+            f"Only url's added to `{allow_url_list}` are accessible. "
+            f"See {help_url} for more information.",
+            translation_domain=DOMAIN,
+            translation_key="remote_url_not_allowed",
+            translation_placeholders={
+                "url": url,
+                "allow_url_list": allow_url_list,
+                "help_url": help_url,
+            },
+        )
+
+    async def _is_valid_file(self, filename: str) -> bool:
+        """Check if a file exists on disk and is in allowlist_external_dirs."""
+
+        file_path = Path(filename).parent
+        if not os.path.exists(file_path) or not os.path.isfile(filename):
+            raise ServiceValidationError(
+                f"Invalid file '{filename}'",
+                translation_domain=DOMAIN,
+                translation_key="invalid_file",
+                translation_placeholders={
+                    "filename": filename,
+                },
+            )
+
+        if self.hass.config.is_allowed_path(filename):
+            return True
+
+        allow_file_list = "allowlist_external_dirs"
+        file_name = os.path.basename(filename)
+        help_url = "https://www.home-assistant.io/docs/configuration/basic/"
+        raise ServiceValidationError(
+            f"Cannot send TvOverlay notification with image file '{file_name}' "
+            f"from directory '{str(file_path)}' which is not secure to load data from. "
+            f"Only folders added to `{allow_file_list}` are accessible. "
+            f"See {help_url} for more information.",
```

Comment:
Sidenote:

We should not have to both provide translation key + placeholders **and** build a very elaborate error message like this.

@jbouwh Why doesn't `HomeAssistantError` generate an error message from the translation by itself?